spec/client_spec.rb

```diff
@@ -115,5 +115,127 @@
         client.address.retrieve('adr_123')
       }.to raise_error(EasyPost::Errors::NotFoundError) # should throw error because our lambda returns 404
     end
+
+    describe 'hooks' do
+      after { EasyPost::Hooks.send(:subscribers).clear }
+
+      it 'subscribes to request events' do
+        notifications = []
+        client = described_class.new(api_key: ENV['EASYPOST_TEST_API_KEY'])
+        client.subscribe_request_hook do |request_data|
+          expect(request_data).to be_an(EasyPost::Hooks::RequestContext)
+          expect(request_data.method).to eq(:get)
+          expect(request_data.path).to end_with('/addresses/adr_123')
+          expect(request_data.headers).to be_a(Hash)
+          expect(request_data.headers['Content-Type']).to eq('application/json')
+          # Because the library is making a GET request, it is expected the request to not have a body
+          expect(request_data.request_body).to be_nil
+          expect(request_data.request_timestamp).to be_a(Time)
+          expect(request_data.request_uuid).to be_a(String)
+          notifications << request_data
+        end
+        expect(EasyPost::Hooks.any_subscribers?(:request)).to eq(true)
+        expect(EasyPost::Hooks.any_subscribers?(:response)).to eq(false)
+
+        expect {
+          client.address.retrieve('adr_123')
+        }.to raise_error(EasyPost::Errors::NotFoundError) # Address doesn't exist
+
+        expect(notifications.size).to eq(1)
+      end
+
+      it 'subscribes to response events' do
+        address_to_create = Fixture.ca_address1
+        lifecycle_request_uuid = ''
+        request_notifications = []
+        response_notifications = []
+        client = described_class.new(api_key: ENV['EASYPOST_TEST_API_KEY'])
+        client.subscribe_request_hook do |request_data|
+          expect(request_data).to be_an(EasyPost::Hooks::RequestContext)
+          expect(request_data.method).to eq(:post)
+          expect(request_data.path).to end_with('/addresses')
+          expect(request_data.headers).to be_a(Hash)
+          expect(request_data.headers['Content-Type']).to eq('application/json')
```

Comment:
nit: asserting that the headers is a hash first before accessing a value of the hash probably isn't necessary since by accessing a value of a hash you are (for free) verifying that the thing in question is a hash.

AKA: If you can access the `Content-Type` of the headers, then it's a hash and so the extra assertion shouldn't be needed.